tests/operators/test_python.py

```diff
@@ -311,6 +312,314 @@ def func(**context):
+
+        self.dag.create_dagrun(
+            run_id='manual__' + DEFAULT_DATE.isoformat(),
+            execution_date=DEFAULT_DATE,
+            start_date=DEFAULT_DATE,
+            state=State.RUNNING
+        )
+        f.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE)
+
+        ds_templated = DEFAULT_DATE.date().isoformat()
+        self.assertEqual(1, len(recorded_calls))
+        self._assert_calls_equal(
+            recorded_calls[0],
+            Call(4,
+                 date(2019, 1, 1),
+                 "dag {} ran on {}.".format(self.dag.dag_id, ds_templated),
+                 Named(ds_templated, 'unchanged'))
+        )
+
+    def test_python_callable_keyword_arguments_are_templatized(self):
+        """Test PythonOperator op_kwargs are templatized"""
+        recorded_calls = []
+
+        task = task_decorator(
+            # a Mock instance cannot be used as a callable function or test fails with a
+            # TypeError: Object of type Mock is not JSON serializable
+            build_recording_function(recorded_calls),
+            dag=self.dag
+        )
+        task(an_int=4, a_date=date(2019, 1, 1), a_templated_string="dag {{dag.dag_id}} ran on {{ds}}.")
+        self.dag.create_dagrun(
+            run_id='manual__' + DEFAULT_DATE.isoformat(),
+            execution_date=DEFAULT_DATE,
+            start_date=DEFAULT_DATE,
+            state=State.RUNNING
+        )
+        task.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE)
+
+        self.assertEqual(1, len(recorded_calls))
+        self._assert_calls_equal(
+            recorded_calls[0],
+            Call(an_int=4,
+                 a_date=date(2019, 1, 1),
+                 a_templated_string="dag {} ran on {}.".format(
+                     self.dag.dag_id, DEFAULT_DATE.date().isoformat()))
+        )
+
+    def test_copy_in_dag(self):
+        with self.dag:
+            self.do_run()
+            self.assertListEqual(['do_run'], self.dag.task_ids)
+            do_run_1 = self.do_run.copy()
+            do_run_2 = self.do_run.copy()
+        self.assertEqual(do_run_1.task_id, 'do_run__1')
+        self.assertEqual(do_run_2.task_id, 'do_run__2')
+
+    def test_copy(self):
+        do_run_1 = self.do_run.copy()
+        do_run_2 = self.do_run.copy()
+        with self.dag:
+            self.do_run()
+            self.assertListEqual(['do_run'], self.dag.task_ids)
+            do_run_1()
+            do_run_2()
+
+        self.assertEqual('do_run__1', do_run_1.task_id)
+        self.assertEqual('do_run__2', do_run_2.task_id)
+
+    def test_dict_outputs(self):
+        @task_decorator(multiple_outputs=True)
+        def return_dict(number: int):
+            return {
+                'number': number + 1,
+                43: 43
+            }
+
+        test_number = 10
+        with self.dag:
+            return_dict(test_number)
+
+        dr = self.dag.create_dagrun(
+            run_id="manual__",
```

Comment:
Please use `DagRunType`

Reply:
Additional info: https://github.com/apache/airflow/pull/8227#issuecomment-626834461

Reply:
Should I rebase on top of your internal branch? Otherwise I can refactor once the PR lands. Leaving this open and adding as todo